x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,273 @@
+                    final CacheFile currentCacheFile = cacheFile.getAndSet(null);
+                    if (currentCacheFile != null) {
+                        currentCacheFile.release(this);
+                    }
+                }
+            }
+        }
+
+        @Override
+        protected void readInternal(final byte[] buffer, final int offset, final int length) throws IOException {
+            final long position = getFilePointer() + this.offset;
+
+            int bytesRead = 0;
+            while (bytesRead < length) {
+                final long pos = position + bytesRead;
+                final int off = offset + bytesRead;
+                final int len = length - bytesRead;
+
+                try {
+                    final CacheFile cacheFile = (parent == null) ? getOrAcquire() : parent.getOrAcquire();
+                    if (cacheFile == null) {
+                        throw new AlreadyClosedException("Failed to acquire a non-evicted cache file");
+                    }
+
+                    try (ReleasableLock ignored = cacheFile.fileLock()) {
+                        bytesRead += cacheFile.fetchRange(pos,
+                            (start, end) -> readCacheFile(cacheFile.getChannel(), end, pos, buffer, off, len),
+                            (start, end) -> writeCacheFile(cacheFile.getChannel(), start, end))
+                            .get();
+                    }
+                } catch (final Exception e) {
+                    if (e instanceof AlreadyClosedException || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
+                        try {
+                            // cache file was evicted during the range fetching, read bytes directly from source
+                            bytesRead += copySource(pos, pos + len, buffer, off);
+                            continue;
+                        } catch (Exception inner) {
+                            e.addSuppressed(inner);
+                        }
+                    }
+                    throw new IOException("Fail to read data from cache", e);
+
+                }
+            }
+            assert bytesRead == length : "partial read operation, read [" + bytesRead + "] bytes of [" + length + "]";
+            assert parent == null || cacheFile.get() == null;
+        }
+
+        int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int offset, long length) throws IOException {
+            assert assertFileChannelOpen(fc);
+            return Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+        }
+
+        @SuppressForbidden(reason = "Use positional writes on purpose")
+        void writeCacheFile(FileChannel fc, long start, long end) throws IOException {
+            assert assertFileChannelOpen(fc);
+            final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(8192L, end - start))];
```

Comment:
I think we should extract `8192L` as a named constant (used here and in `copySource`).

Reply:
Sure, I added `COPY_BUFFER_SIZE` in https://github.com/elastic/elasticsearch/pull/50693/commits/99a99686fe426daa450742b9ac1c33f966978bc6